[자판기 앱 Step1 : 아이패드 앱] 미션 제출합니다.

ehgud0670 opened the pull request:
> 상속과 관련하여 고민한 점 
```
우유 - 딸기우유, 초코우유, 바나나우유 택 1
탄산음료 - 콜라, 사이다, 환타 택 1
커피 - TOP, 칸타타, 조지아 택 1
```
=> lucas 에 나와있는 예시인데 예시를 보고 헷갈렸던 점이 있습니다.
커피는 브랜드명(TOP,칸타타,조지아를 브랜드라고 판단함)인데 , 
우유나 탄산음료는 해당 음료의 종류라서 결이 달라보였습니다. 

그래서 저는 어떻게 상속을 할지 고민해보고,,,

```
Beverage (1단계)
Milk, Soda, Coffee (2단계)
HersheyChocolateDrink / Pepsi / Cantata (3단계)
```
위의 보이는 것처럼 **일관성을 위해** 3단계의 음료 종류를 모두 브랜드명으로 두었습니다.  

또,

* 허쉬초콜릿드링크의 음료 종류: 오리지날, 밀크 초콜릿, 초콜릿, 쿠키 앤 크림
* 칸타타의 음료 종류 : 프리미엄 라떼, 아메리카노, 더치블랙, 카라멜 마키아토 등등
* 펩시의 음료 종류 : 펩시 콜라, 다이어트 콜라 등등

위에 보이는 것처럼 브랜드에 따라 여러가지 이름의 음료 제품이 존재할수 있기 때문에 
각각 브랜드에 따라 **여러 다른 객체를 만들수 있다고 판단하였고**, 클래스 이름을 브랜드 이름으로 두었습니다.

ehgud0670 committed:
Setup: 프로젝트 기본 설정함
         * 프로젝트에 대한 일반 설정 (General)에서 Devices 항목을 iPad로 변경하고, Device Orientation 항목을 가로 (Landscape Left, Landscape Right)만 선택
         * 스토리보드에서 ipad pro 10.5로 보이게 설정함

ehgud0670 committed:
Feat: 자판기 구조체 추가

ehgud0670 committed:
Feat: 최상위 클래스인 Beverage 클래스 코드 작성

ehgud0670 committed:
Feat: 최상위 객체인 Beverage를 상속하는 상위 클래스들 코드 작성함
	Milk: 우유를 나타내는 객체
	Coffee: 커피를 나타내는 객체
	Soda: 탄산음료를 나타내는 객체

ehgud0670 committed:
Feat: Coffee 클래스를 상속하는 Cantata 클래스 구현
      1. 회사는 "롯데칠성음료"가 확실하므로 convenience init을 이용하여
	 default value로 선정함
      2. 나머지 프로퍼티들은 무슨 칸타타 음료에 따라서 다 다른 값이므로
	 생성자에서 매개변수로 받게끔 코드 구현함

ehgud0670 committed:
Feat: Soda 클래스를 상속하는 Pepsi 클래스 구현
          1. 회사는 "펩시코"가 확실하므로 convenience init을 이용하여
             default value로 선정함
          2. 나머지 프로퍼티들은 무슨 펩시 음료냐에 따라서 다 다른 값이므로
             생성자에서 매개변수로 받게끔 코드 구현

ehgud0670 committed:
Feat: Milk 클래스를 상속하는 HersheyChocolateDrink 클래스 구현
      1. 회사는 "매일유업"가 확실하므로 convenience init을 이용하여
         default value로 선정함
      2. 나머지 프로퍼티들은 무슨 허쉬초콜릿드링크 음료냐에 따라서 다 다른 값이므로
         생성자에서 매개변수로 받게끔 코드 구현

ehgud0670 committed:
Feat: VendingMachine 구조체를 프로퍼티로 추가, let으로 선언
     이유: let으로 일단 선언하고, 이후 step에서 프로퍼티를 바꿔야 한다면
	  var 으로 재선언 할 것이다.

ehgud0670 committed:
Refactor: init의 날짜 매개변수를 String 타입으로 수정함
           이유 : 생성할때 편하게 날짜를 나타내는 문자열을 보내려고
           해당 값이 오류가 날 문자열인 경우
           => 해당 문자열이 날짜 형식이 아닐 경우 init이 안되게끔 init?을 사용함

           추가적으로 Beverage class 에 날짜 형식을 나타내는 DateFormatter도
           정적(static) 프로퍼티로 둠
	   => 효과 1: 음료 객체 생성할때 DateFormatter가 설정한 날짜 형식이 아니면 생성이 안된다.
           => 효과 2: 음료 객체 클래스 출력할때 날짜 출력을 형식에 맞게 출력할 수 있다.

ehgud0670 committed:
Refactor: 브랜드 default 값 재수정함.
	이유: 이전에 브랜드 명을 제조하는 회사명으로 착각함
	      따라서 다시 브랜드 명을 수정함("허쉬","칸타타","펩시")

ehgud0670 committed:
Style: 클래스 선언부와 메소드 사이 개행을 띄워줌

ehgud0670 committed:
Refactor: 브랜드 명을 재수정함. 클래스 명과 브랜드명은 일치해야 한다.

	이유: 1. 허쉬초콜릿드링크의 음료 종류: 오리지날, 밀크 초콜릿, 초콜릿, 쿠키 앤 크림
 	      2. 칸타타의 음료 종류 : 프리미엄 라떼, 아메리카노, 더치블랙, 카라멜 마키아토 등등
	      3. 펩시의 음료 종류 : 펩시 콜라, 다이어트 콜라 등등

	=> 위에 보이는 것처럼 브랜드에 따라 여러가지 이름의 음료 제품이 존재한다.
	   따라서 각각 브랜드에 따라 여러 다른 객체를 만들수 있다고 판단,
	   클래스 이름을 브랜드 이름으로 두었다.

ehgud0670 committed:
Feat: viewDidLoad에 음료 객체 3가지 추가하고 음료 객체의 내용을 print함

ehgud0670 committed:
Docs: 1단계와 관련한 사항들 작성함

ehgud0670 committed:
Docs: 허쉬 철자 틀린것 수정함

godrm commented:
머지할 브랜치를 master에서 변경해주세요.

godrm commented:
> 위의 보이는 것처럼 일관성을 위해 3단계의 음료 종류를 모두 브랜드명으로 두었습니다.

잘하셨네요. 클래스 계층을 구분하는 기준을 명확하게 나누는 게 도움이 될 겁니다.

godrm commented on VendingMachineApp/VendingMachineApp/View/Beverage.swift:
```diff
@@ -0,0 +1,45 @@
+        formatter.locale = Locale(identifier:"ko_KR")
+        formatter.dateFormat = "yyyyMMdd"
+        return formatter
+    }()
```
날짜를 문자열로 변경하는 것은 음료수 객체가 할 책임이 아닌 것 같습니다. 
다른 어떤 객체가 포함하는 게 좋을까요?

godrm commented on VendingMachineApp/VendingMachineApp/View/Cantata.swift:
```diff
@@ -0,0 +1,24 @@
+                  name: String, dateInfo: String) {
+        super.init(brand: brand, volume: volume, price: price,
+                   name: name, dateInfo: dateInfo)
+    }
```
칸타타를 강제해야 한다면 지정 생성자를 안 만들고 아래 conveience를 지정 생성자로 만드는 것도 방법이 되겠네요.

ehgud0670 replied:
음 넵, convenience init만이 default value을 설정할수 있다고 생각해서 이렇게 작성한 것 같습니다. 
말씀하신 방법이 훨씬 적절하겠네요 반영하겠습니다.  강제해야 되서 되게 좋습니다.👍👍

godrm commented on VendingMachineApp/VendingMachineApp/View/Coffee.swift:
```diff
@@ -0,0 +1,19 @@
+
+class Coffee: Beverage {
+    
+    override init?(brand: String, volume: Int, price: Int,
```
하위 클래스를 만들었는데 부모 클래스와 차이점이 전혀 없다면 의미가 없는 겁니다. 그냥 부모 클래스를 써도 된다는 거니까요
세부적인 속성을 갖고, 기능이 달라져야겠죠.

ehgud0670 replied:
넵, 사실 다음 step2를 보고 3단계로 상속했습니다. 
다음 미션에서 세부적인 속성들을 만들어보도록 하겠습니다. 감사합니다. 👍👍👍

godrm commented on VendingMachineApp/VendingMachineApp/View/VendingMachine.swift:
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+struct VendingMachine {
+    
```
아직 구현하지 않은 거라면 커밋하지 않아도 될 것 같습니다.

ehgud0670 replied:
넵, 다음부턴 커밋하지 않도록 주의하겠습니다. 감사합니다. 👍👍👍

godrm reviewed (commented):
클래스 설계에 기준을 정리해서 좋네요. 
다만 상속 관계를 표현할 때는 세부적으로 다른 것을 기준으로 정하는 게 좋습니다. 피드백을 참고하세요.

ehgud0670 committed:
Refactor: 기존의 override init과 convenience init 을 모두 삭제하고 지정자 init을 새로 생성함
	  장점 : 새로 만든 지정자 init으로 브랜드를 강제할 수 있다.
		 따라서 개발자가 잘못 객체를 만들 가능성이 훨씬 줄어든다.

ehgud0670 committed:
Refactor: Util 클래스를 만들어서 dateFormatter로 날짜를 문자열로 바꾸게 하는 기능을 하게함
	이유 : 음료 객체는 날짜를 문자열로 만들 책임 및 역할이 아니다.
	       따라서 전반적으로 쓰이는 곳을 말하는 Util이라는 클래스를 생성해서
	       Util에서 해당 역할을 하도록 하였다.

ehgud0670 committed:
Style: 폴더 이름 Views로 바꿈

ehgud0670 committed:
Style: 선언 부분에 한칸씩 띄어지도록 코드 수정

ehgud0670 committed:
Style: 프로퍼티 선언부분 일관성을 위해 공백 지움

ehgud0670 committed:
Style: 잘못된 개행 바로 잡음

ehgud0670 commented:
* 브랜치 ehgud0670으로 변경하였습니다. 
* 피드백 모두 반영하였습니다. 

> 궁금한 점

날짜를 문자열로 변경하는 역할은 새로 Util이라는 클래스를 만들고 static한 dateFormatter 속성이 하도록 변경하였습니다. 
이 역할은 음료 객체가 할 역할이 아니므로 전의 코드보다 더 나아진 것 같은데,, 
매번 Util에만 이런 역할은 주는 것 같아 뭔가 아쉽네용.. 더 좋은 방법이 있을까요?? 🤷‍♂️

godrm commented:
Util 객체가 그런걸 모아놓는 경우가 있지만
스위프트에서는 이미 만들어놓은 객체를 extension 할 수 있죠.

ehgud0670 commented:
넵 그럼 extension을 활용해보겠습니다.
